### PR TITLE
Heap preinitialization with thread level support

### PR DESCRIPTION
**Pull request by wrrobin**

### CODE REVIEW

File: mpp/shmemx_c_func.h4 [hunk 1]
```diff
@@ -38,4 +38,5 @@ SHMEM_FUNCTION_ATTRIBUTES void SHPRE()shmemx_pcntr_get_all(shmem_ctx_t ctx, shme
 /* Separate initializers */
 SHMEM_FUNCTION_ATTRIBUTES void SHPRE()shmemx_heap_create(void *base, size_t size, int device_type, int device_index);
 SHMEM_FUNCTION_ATTRIBUTES void SHPRE()shmemx_heap_preinit();
+SHMEM_FUNCTION_ATTRIBUTES int SHPRE()shmemx_heap_preinit_thread(int requested, int *provided);
 SHMEM_FUNCTION_ATTRIBUTES void SHPRE()shmemx_heap_postinit();
```
Comment (davidozog) [inline mpp/shmemx_c_func.h4:42]: @wrrobin - do we want weak symbols for these (for pshmem)?

**Review by davidozog — APPROVED** [review]

**Comment by philipmarshall21** [comment]
Sorry what is meant by "thread level support" in this context? Is the new heap preinitialization function needed when SOS is configured with some of the various threading options (**SHMEM_THREAD_SINGLE**/**SHMEM_THREAD_FUNNELED**/**SHMEM_THREAD_SERIALIZED**/etc.)?

File: mpp/shmemx_c_func.h4 (hunk 1 above)
Comment (wrrobin) [reply mpp/shmemx_c_func.h4:42]: Oh.. yes. Thanks. I will add..

**Comment by wrrobin** [comment]
> Sorry what is meant by "thread level support" in this context? Is the new heap preinitialization function needed when SOS is configured with some of the various threading options (**SHMEM_THREAD_SINGLE**/**SHMEM_THREAD_FUNNELED**/**SHMEM_THREAD_SERIALIZED**/etc.)?

SOS can be initialized with different thread levels as you pointed out. Our previous heap preinit only supports THREAD_SINGLE. The new one will enable other thread options.

**Review by philipmarshall21 — APPROVED** [review]

**Pull request merged**